### PR TITLE
etcdmain: gateway monitor-interval flag

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/5335.

### CODE REVIEW

**Comment by gyuho** [comment]
@xiang90 Plus, changed flag type to string slice, so we don't need to parse it ourselves.

**Comment by xiang90** [comment]
Can we add warning logs to proxy when we mark an endpoint to be available? and also we shall add notice logs to proxy when we mark them back to be available.

File: etcdmain/gateway.go [hunk 1]
```diff
@@ -60,23 +61,23 @@ func newGatewayStartCommand() *cobra.Command {
 	}
 
 	cmd.Flags().StringVar(&gatewayListenAddr, "listen-addr", "127.0.0.1:23790", "listen address")
-	cmd.Flags().StringVar(&gatewayEndpoints, "endpoints", "127.0.0.1:2379", "comma separated etcd cluster endpoints")
+	cmd.Flags().StringSliceVar(&gatewayEndpoints, "endpoints", []string{"127.0.0.1:2379"}, "comma separated etcd cluster endpoints")
+	cmd.Flags().DurationVar(&getewayMonitorInterval, "monitor-interval", 5*time.Minute, "interval to monitor failed endpoints for retrying")
```
Comment (xiang90) [inline etcdmain/gateway.go]: change default to 1 minute?

**Comment by gyuho** [comment]
@xiang90 Addressed. PTAL.

Thanks.

File: proxy/tcpproxy/userspace.go [hunk 2]
```diff
@@ -31,6 +37,7 @@ func (r *remote) inactivate() {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 	r.inactive = true
+	plog.Noticef("inactivated %s", r.addr)
```
Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: i think this should be warning as i mentioned. also we need to log why it is inactive (the error) and how long will we retry on this remote?

File: etcdmain/gateway.go [hunk 3]
```diff
@@ -60,23 +61,23 @@ func newGatewayStartCommand() *cobra.Command {
 	}
 
 	cmd.Flags().StringVar(&gatewayListenAddr, "listen-addr", "127.0.0.1:23790", "listen address")
-	cmd.Flags().StringVar(&gatewayEndpoints, "endpoints", "127.0.0.1:2379", "comma separated etcd cluster endpoints")
+	cmd.Flags().StringSliceVar(&gatewayEndpoints, "endpoints", []string{"127.0.0.1:2379"}, "comma separated etcd cluster endpoints")
+	cmd.Flags().DurationVar(&getewayMonitorInterval, "monitor-interval", time.Minute, "interval to monitor failed endpoints for retrying")
```
Comment (xiang90) [inline etcdmain/gateway.go]: failure-retry-interval?
/cc @heyitsanthony need your help on naming...

File: proxy/tcpproxy/userspace.go (hunk 2 above)
Comment (gyuho) [inline proxy/tcpproxy/userspace.go]: Right. Will fix.

File: etcdmain/gateway.go (hunk 3 above)
Comment (heyitsanthony) [inline etcdmain/gateway.go]: `retry-delay` like curl?

**Comment by gyuho** [comment]
@xiang90 @heyitsanthony Changed flag name to `retry-delay`, and added more logs.

PTAL.

Thanks.

File: etcdmain/gateway.go [hunk 4]
```diff
@@ -60,23 +61,23 @@ func newGatewayStartCommand() *cobra.Command {
 	}
 
 	cmd.Flags().StringVar(&gatewayListenAddr, "listen-addr", "127.0.0.1:23790", "listen address")
-	cmd.Flags().StringVar(&gatewayEndpoints, "endpoints", "127.0.0.1:2379", "comma separated etcd cluster endpoints")
+	cmd.Flags().StringSliceVar(&gatewayEndpoints, "endpoints", []string{"127.0.0.1:2379"}, "comma separated etcd cluster endpoints")
+	cmd.Flags().DurationVar(&getewayRetryDelay, "retry-delay", time.Minute, "interval to retry monitoring failed endpoints")
```
Comment (xiang90) [inline etcdmain/gateway.go]: also fix the short message for the flag?

File: proxy/tcpproxy/userspace.go [hunk 5]
```diff
@@ -31,6 +37,7 @@ func (r *remote) inactivate() {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 	r.inactive = true
+	plog.Warningf("inactivated %s", r.addr)
```
Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: remove this? we already have same thing printed at line 116

File: proxy/tcpproxy/userspace.go [hunk 6]
```diff
@@ -105,6 +113,7 @@ func (tp *TCPProxy) serve(in net.Conn) {
 		if err == nil {
 			break
 		}
+		plog.Warningf("inactivating %s with error %v", remote.addr, err)
```
Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: print out how long will it be retried.

**Comment by gyuho** [comment]
@xiang90 All addressed. PTAL.

Thanks.

File: proxy/tcpproxy/userspace.go [hunk 7]
```diff
@@ -105,6 +112,7 @@ func (tp *TCPProxy) serve(in net.Conn) {
 		if err == nil {
 			break
 		}
+		plog.Warningf("inactivating %s with error %v (retry in %v)", remote.addr, err, tp.MonitorInterval)
```
Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: move this to line 117

deactivated endpoint [%s] due to %v for %v.

File: proxy/tcpproxy/userspace.go [hunk 8]
```diff
@@ -42,6 +48,7 @@ func (r *remote) tryReactivate() {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 	r.inactive = false
+	plog.Printf("activated %s", r.addr)
```
Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: we also need to print out something at line 44.

**Comment by gyuho** [comment]
@xiang90 Just fixed the log and add another log for error at line 44.

PTAL.

Thanks

File: proxy/tcpproxy/userspace.go [hunk 9]
```diff
@@ -36,12 +42,14 @@ func (r *remote) inactivate() {
 func (r *remote) tryReactivate() {
 	conn, err := net.Dial("tcp", r.addr)
 	if err != nil {
+		plog.Warningf("tryReactivate failed to connect to endpoint [%s] due to %v", r.addr, err)
```
Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: this will make the remote to stay inactive for another interval. we should also log that information.

Comment (gyuho) [inline proxy/tcpproxy/userspace.go]: Yeah right. Will be more specific.

**Comment by gyuho** [comment]
@xiang90 Just fixed. PTAL.

File: proxy/tcpproxy/userspace.go [hunk 10]
```diff
@@ -33,15 +39,17 @@ func (r *remote) inactivate() {
 	r.inactive = true
 }
 
-func (r *remote) tryReactivate() {
+func (r *remote) tryReactivate(it time.Duration) {
```
Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: probably we should return error from tryReactive and print out everything in monitor function. it is unclear why this function needs an additional it argument.

Comment (gyuho) [inline proxy/tcpproxy/userspace.go]: Ok will fix.

File: proxy/tcpproxy/userspace.go [hunk 11]
```diff
@@ -141,7 +150,11 @@ func (tp *TCPProxy) runMonitor() {
 			tp.mu.Lock()
 			for _, r := range tp.remotes {
 				if !r.isActive() {
-					go r.tryReactivate()
```
Comment (xiang90) [inline proxy/tcpproxy/userspace.go]:
```go
go func() {
    err := r.tryReactive()
    ...
}
```

**Comment by gyuho** [comment]
@xiang90 Fixed more, to print logs inside `runMonitor`. PTAL.

Will fix more at home, if you have more feedback.

Thanks.

File: proxy/tcpproxy/userspace.go (hunk 11 above)
Comment (gyuho) [inline proxy/tcpproxy/userspace.go]: ok will fix.

File: proxy/tcpproxy/userspace.go [hunk 12]
```diff
@@ -33,16 +39,18 @@ func (r *remote) inactivate() {
 	r.inactive = true
 }
 
-func (r *remote) tryReactivate() {
+func (r *remote) tryReactivate() error {
```
Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: please change this back. the blocking point is net.Dial. we need to ensure the whole func is unblocking.

Comment (gyuho) [inline proxy/tcpproxy/userspace.go]: @xiang90 You are right. Just changed it back and wrap the whole function in goroutine as you suggested. PTAL.

File: proxy/tcpproxy/userspace.go [hunk 13]
```diff
@@ -141,7 +148,13 @@ func (tp *TCPProxy) runMonitor() {
 			tp.mu.Lock()
 			for _, r := range tp.remotes {
 				if !r.isActive() {
-					go r.tryReactivate()
+					go func() {
+						if err := r.tryReactivate(); err != nil {
+							plog.Warningf("failed to connect to endpoint [%s] due to %v (stay inactive for another %v)", r.addr, err, tp.MonitorInterval)
```
Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: failed to active endpoint [%s] due to %v (...)

**Comment by xiang90** [comment]
LGTM after fixing the last issue.

**Comment by gyuho** [comment]
Just fixed. Thanks for all those feedback. Merging in greens.

**Pull request merged**